File: src/main/java/lekkit/scev/client/screen/MachineScreen.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.client.screen;

import com.mojang.blaze3d.systems.RenderSystem;
import lekkit.rvvm.HIDKeyboard;
import lekkit.scev.client.DisplayManager;
import lekkit.scev.client.DisplayState;
import lekkit.scev.menu.MachineMenu;
import lekkit.scev.network.MachineInputPayload;
import lekkit.scev.network.MachineResetPayload;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.inventory.Slot;
import net.neoforged.neoforge.network.PacketDistributor;
import org.lwjgl.glfw.GLFW;

/**
 * Screen that displays a running machine's framebuffer. Keyboard / mouse input is
 * converted to {@link MachineInputPayload}s and sent to the server.
 *
 * <p><b>Sizing.</b> The native framebuffer is {@value #SCREEN_W}×{@value #SCREEN_H}.
 * Rendering it 1:1 into Minecraft's GUI space produced a 672×512 window (plus
 * margin) that clipped on most real-world display setups — on a 1920×1080
 * window with GUI scale 2, the GUI space is 960×540 and the window didn't fit
 * vertically; at scale 3 (GUI space 640×360) it didn't fit in either axis.
 *
 * <p>We now compute a scale factor at {@link #init()} to fit the framebuffer
 * inside the current GUI space with a small margin, preserving the 4:3
 * aspect ratio. Mouse coordinates passed to the server are un-scaled so the
 * VM still receives framebuffer-space pixel positions.
 *
 * <p><b>Held-key tracking.</b> We keep the set of HID keys we've pressed
 * (without release). On screen close / focus lost, every tracked key is
 * released — so if the player closes the GUI while holding shift, the VM
 * doesn't see shift as stuck down forever.
 */
public class MachineScreen extends AbstractContainerScreen<MachineMenu> {
    /** Pixel size of the framebuffer — matches Framebuffer(640,480). */
    private static final int SCREEN_W = 640;
    private static final int SCREEN_H = 480;
    /** Space reserved around the framebuffer inside the GUI window. */
    private static final int SCREEN_MARGIN = 16;
    /** Extra vertical space below the framebuffer for the power/reset buttons. */
    private static final int BUTTON_STRIP_H = 16;

    /**
     * Cap on how much clipboard text a single Ctrl+Shift+V will type. Past
     * this, the user is pasting a file — which they should do via a guest-
     * side transfer, not synthetic keystrokes. 4096 chars at
     * {@link ClipboardPaster#DEFAULT_EVENTS_PER_TICK}=4 ev/tick / 20 tps =
     * ~205 s of typing, which is already generous.
     */
    private static final int MAX_PASTE_CHARS = 4096;

    /** Modifier HIDs we suppress before typing pasted text so the guest doesn't see Ctrl+letter etc. */
    private static final byte[] MODIFIER_HIDS = {
            HIDKeyboard.HID_KEY_LEFTCTRL, HIDKeyboard.HID_KEY_LEFTSHIFT,
            HIDKeyboard.HID_KEY_LEFTALT, HIDKeyboard.HID_KEY_LEFTMETA,
            HIDKeyboard.HID_KEY_RIGHTCTRL, HIDKeyboard.HID_KEY_RIGHTSHIFT,
            HIDKeyboard.HID_KEY_RIGHTALT, HIDKeyboard.HID_KEY_RIGHTMETA,
    };

    /** Tracks keys we've emitted KEY_PRESS for but not yet KEY_RELEASE. */
    private final HeldKeyTracker heldKeys = new HeldKeyTracker(
            hid -> PacketDistributor.sendToServer(MachineInputPayload.keyPress(hid)),
            hid -> PacketDistributor.sendToServer(MachineInputPayload.keyRelease(hid)));

    /**
     * Paces clipboard paste as synthetic keystrokes. Same press/release
     * sinks as {@link #heldKeys}, but events here aren't tracked as "held"
     * because they always come in matched press/release pairs inside a
     * single paste operation.
     */
    private final ClipboardPaster paster = new ClipboardPaster(
            hid -> PacketDistributor.sendToServer(MachineInputPayload.keyPress(hid)),
            hid -> PacketDistributor.sendToServer(MachineInputPayload.keyRelease(hid)));

    /** Computed once in {@link #init()}. 1.0 = render at native 640×480 resolution. */
    private float displayScale = 1.0f;
    /** Rendered framebuffer size in GUI pixels ({@code SCREEN_W * displayScale}, rounded). */
    private int displayW = SCREEN_W;
    private int displayH = SCREEN_H;
    /** Top-left corner of the rendered framebuffer in GUI space. */
    private int displayX;
    private int displayY;

    public MachineScreen(MachineMenu menu, Inventory inv, Component title) {
        super(menu, inv, title);
        // imageWidth/imageHeight are used by AbstractContainerScreen to position
        // leftPos/topPos. We set them provisionally here; init() recomputes the
        // framebuffer size + position based on the actual window, because we
        // don't know `width` / `height` yet in the constructor.
        this.imageWidth = SCREEN_W + 2 * SCREEN_MARGIN;
        this.imageHeight = SCREEN_H + 2 * SCREEN_MARGIN + BUTTON_STRIP_H;
    }

    @Override
    protected void init() {
        // Compute a scale that fits the framebuffer (plus margin + buttons) into
        // the current GUI space. We never upscale — small GUI scales keep the
        // native 640×480 resolution. For larger GUI scales (smaller GUI space)
        // we shrink while preserving the 4:3 aspect.
        int availW = Math.max(SCREEN_W / 4, this.width - 2 * SCREEN_MARGIN);
        int availH = Math.max(SCREEN_H / 4, this.height - 2 * SCREEN_MARGIN - BUTTON_STRIP_H);
        float scaleX = availW / (float) SCREEN_W;
        float scaleY = availH / (float) SCREEN_H;
        this.displayScale = Math.min(1.0f, Math.min(scaleX, scaleY));
        this.displayW = Math.max(1, Math.round(SCREEN_W * displayScale));
        this.displayH = Math.max(1, Math.round(SCREEN_H * displayScale));

        // Size the AbstractContainerScreen "image" box to the actual display
        // area (plus button strip). leftPos / topPos are centred on this.
        this.imageWidth = displayW + 2 * SCREEN_MARGIN;
        this.imageHeight = displayH + 2 * SCREEN_MARGIN + BUTTON_STRIP_H;

        super.init();

        this.displayX = leftPos + SCREEN_MARGIN;
        this.displayY = topPos + SCREEN_MARGIN;

        // Single power button at the top-right of the framebuffer. Off→on
        // reboots from bootrom; separate warm-reset affordance removed — see
        // ComputerCaseScreen javadoc for rationale.
        int btnX = displayX + displayW - 34;
        int btnY = topPos + imageHeight - BUTTON_STRIP_H + 2;
        addRenderableWidget(Button.builder(Component.translatable("button.scev.power"),
                b -> PacketDistributor.sendToServer(new MachineResetPayload(false)))
                .bounds(btnX, btnY, 32, 12).build());

        // Ensure no widget is holding keyboard focus — all key events should
        // reach our keyPressed / keyReleased so they're forwarded to the VM,
        // not swallowed by a button waiting for ENTER/SPACE.
        this.setFocused(null);
    }

    @Override
    protected void renderBg(GuiGraphics g, float partialTick, int mouseX, int mouseY) {
        DisplayState display = DisplayManager.get(menu.getMachineUuid());
        if (display == null) {
            // VM not running (never powered on, or powered off while the
            // screen stayed open). Paint a solid black rect so the display
            // area doesn't fall through to the dimmed-world layer behind —
            // players read that as "my screen vanished" rather than "the VM
            // is off."
            g.fill(displayX, displayY, displayX + displayW, displayY + displayH, 0xFF000000);
            return;
        }
        ResourceLocation tex = display.getOrUploadTexture();
        RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
        // Scale the native 640×480 texture into the computed display area.
        // blit(tex, x, y, destW, destH, u, v, srcW, srcH, texW, texH).
        g.blit(tex, displayX, displayY, displayW, displayH, 0, 0, SCREEN_W, SCREEN_H, SCREEN_W, SCREEN_H);
    }

    @Override
    public void render(GuiGraphics g, int mouseX, int mouseY, float partialTick) {
        renderBackground(g, mouseX, mouseY, partialTick);
        super.render(g, mouseX, mouseY, partialTick);
    }

    /**
     * Suppress the default slot rendering. The menu's player-inventory slots
     * exist only so that shift-click from another container doesn't crash —
     * they don't correspond to any real container on the server side, so we
     * hide them visually rather than drawing them on top of the framebuffer.
     */
    @Override
    protected void renderSlot(GuiGraphics g, Slot slot) {
        // no-op: framebuffer view doesn't have slots
    }

    /** Hide the default "Inventory" / title labels — they clutter the framebuffer view. */
    @Override
    protected void renderLabels(GuiGraphics g, int mouseX, int mouseY) {
        // no-op
    }

    @Override public boolean isPauseScreen() { return false; }

    /**
     * Client tick hook — fires ~20 times/s. We use it to drain a small
     * batch of pending paste events, so a large clipboard doesn't burst
     * past the guest's HID buffer capacity in a single frame.
     */
    @Override
    public void containerTick() {
        super.containerTick();
        paster.tick();
    }

    /* ---------------- Input: keyboard ---------------- */

    @Override
    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
        // Leave ESCAPE alone so the player can close the screen.
        if (keyCode == GLFW.GLFW_KEY_ESCAPE) return super.keyPressed(keyCode, scanCode, modifiers);

        // Ctrl+Shift+V (or Cmd+Shift+V on macOS) pastes the host clipboard as
        // synthetic keystrokes. Plain Ctrl+V still flows through to the guest
        // — many guest programs use Ctrl+V themselves, so we only claim the
        // three-modifier combo. The V key itself is consumed (not tracked as
        // held) so the guest doesn't see a stray 'v' before the paste stream.
        if (keyCode == GLFW.GLFW_KEY_V
                && (modifiers & GLFW.GLFW_MOD_SHIFT) != 0
                && (modifiers & (GLFW.GLFW_MOD_CONTROL | GLFW.GLFW_MOD_SUPER)) != 0) {
            triggerPaste();
            return true;
        }

        byte hid = GlfwToHid.map(keyCode);
        if (hid != 0) {
            heldKeys.press(hid);
            return true;
        }
        return super.keyPressed(keyCode, scanCode, modifiers);
    }

    /**
     * Read the host clipboard and queue it for typing into the guest. We
     * first emit release events for any modifier keys the tracker currently
     * thinks are held (the Ctrl+Shift that triggered us, at minimum) so the
     * guest doesn't interpret the pasted letters as hotkey chords.
     *
     * <p>We don't re-press those modifiers afterwards. If the player is
     * still holding Ctrl+Shift when paste finishes, GLFW will eventually
     * fire a release we'll forward normally — the VM sees a release-of-
     * unheld key, which HID handles as a no-op. The tradeoff is that any
     * keystrokes the player types between paste-end and their physical
     * release won't carry the original modifiers; in practice nobody does
     * that within the short paste window.
     */
    private void triggerPaste() {
        String text = Minecraft.getInstance().keyboardHandler.getClipboard();
        if (text == null || text.isEmpty()) return;
        if (text.length() > MAX_PASTE_CHARS) text = text.substring(0, MAX_PASTE_CHARS);

        for (byte mod : MODIFIER_HIDS) {
            if (heldKeys.isHeld(mod)) {
                // Release on the wire but leave the tracker's entry in place.
                // When the player physically releases the key, tracker.release
                // will forward another release — harmless to the guest.
                paster.queueRelease(mod);
            }
        }
        paster.queueText(text);
    }

    @Override
    public boolean keyReleased(int keyCode, int scanCode, int modifiers) {
        // Don't emit a release for ESCAPE — we never emitted a press for it
        // (see keyPressed above), and the escape that closes the screen was
        // pressed *before* the screen opened so the VM didn't see it either.
        if (keyCode == GLFW.GLFW_KEY_ESCAPE) return super.keyReleased(keyCode, scanCode, modifiers);

        byte hid = GlfwToHid.map(keyCode);
        if (hid != 0) {
            heldKeys.release(hid);
            return true;
        }
        return super.keyReleased(keyCode, scanCode, modifiers);
    }

    /**
     * Consume {@code charTyped} events for any key we would have handled in
     * keyPressed (letters, numbers, symbols). Otherwise AbstractContainerScreen's
     * default charTyped may leak text into debug tools or kick focus around —
     * the VM has already received the HID scancode via keyPressed, and there's
     * no text field to type into in this screen.
     */
    @Override
    public boolean charTyped(char codePoint, int modifiers) {
        return true;
    }

    /* ---------------- Input: mouse ---------------- */

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        if (isInScreenRect(mouseX, mouseY)) {
            byte hidBtn = mapMouseButton(button);
            if (hidBtn != 0) {
                PacketDistributor.sendToServer(MachineInputPayload.mousePress(hidBtn));
                emitMousePlace(mouseX, mouseY);
                return true;
            }
        }
        return super.mouseClicked(mouseX, mouseY, button);
    }

    @Override
    public boolean mouseReleased(double mouseX, double mouseY, int button) {
        byte hidBtn = mapMouseButton(button);
        if (hidBtn != 0) {
            PacketDistributor.sendToServer(MachineInputPayload.mouseRelease(hidBtn));
            return true;
        }
        return super.mouseReleased(mouseX, mouseY, button);
    }

    @Override
    public boolean mouseDragged(double mouseX, double mouseY, int button, double dx, double dy) {
        if (isInScreenRect(mouseX, mouseY)) {
            emitMousePlace(mouseX, mouseY);
            return true;
        }
        return super.mouseDragged(mouseX, mouseY, button, dx, dy);
    }

    @Override
    public boolean mouseScrolled(double mouseX, double mouseY, double scrollX, double scrollY) {
        if (isInScreenRect(mouseX, mouseY) && scrollY != 0) {
            byte delta = (byte) (scrollY > 0 ? 1 : -1);
            PacketDistributor.sendToServer(MachineInputPayload.mouseScroll(delta));
            return true;
        }
        return super.mouseScrolled(mouseX, mouseY, scrollX, scrollY);
    }

    @Override
    public void mouseMoved(double mouseX, double mouseY) {
        if (isInScreenRect(mouseX, mouseY)) {
            emitMousePlace(mouseX, mouseY);
        }
        super.mouseMoved(mouseX, mouseY);
    }

    /* ---------------- Lifecycle ---------------- */

    @Override
    public void removed() {
        // Drop any paste events that hadn't been drained yet — they'd arrive
        // at the guest without the original paste context and potentially
        // land in the wrong window / tty. Better to lose the tail than to
        // scatter random characters.
        paster.clear();
        // Release any keys we've left pressed. Otherwise the VM sees them as
        // stuck down (e.g. shift jammed on) forever. Done through the tracker
        // so HeldKeyTrackerTest can verify the contract without a real screen.
        heldKeys.releaseAll();
        super.removed();
    }

    /* ---------------- Helpers ---------------- */

    private boolean isInScreenRect(double mouseX, double mouseY) {
        return mouseX >= displayX && mouseX < displayX + displayW
                && mouseY >= displayY && mouseY < displayY + displayH;
    }

    private void emitMousePlace(double mouseX, double mouseY) {
        // Map GUI-space coords back to framebuffer pixels, accounting for scale.
        float scaleInv = displayScale == 0f ? 1f : 1f / displayScale;
        int x = (int) Math.max(0, Math.min(SCREEN_W - 1, (mouseX - displayX) * scaleInv));
        int y = (int) Math.max(0, Math.min(SCREEN_H - 1, (mouseY - displayY) * scaleInv));
        PacketDistributor.sendToServer(MachineInputPayload.mousePlace((short) x, (short) y));
    }

    /**
     * Map GLFW mouse button to HID mouse button bit. HID buttons: 1=left,
     * 2=right, 4=middle (bitmask). Our payload carries a single byte — we
     * re-use that as "which bit to toggle" so the server handler sets/clears
     * the right bit on the HID mouse state.
     */
    private static byte mapMouseButton(int glfwButton) {
        return switch (glfwButton) {
            case GLFW.GLFW_MOUSE_BUTTON_LEFT -> 1;
            case GLFW.GLFW_MOUSE_BUTTON_RIGHT -> 2;
            case GLFW.GLFW_MOUSE_BUTTON_MIDDLE -> 4;
            default -> 0;
        };
    }
}

File: src/main/kotlin/lekkit/scev/client/screen/MachineScreen.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.client.screen

import com.mojang.blaze3d.systems.RenderSystem
import io.wispforest.owo.ui.component.Components
import io.wispforest.owo.ui.container.FlowLayout
import io.wispforest.owo.ui.core.Sizing
import io.wispforest.owo.ui.core.Surface
import java.util.function.Consumer
import lekkit.rvvm.HIDKeyboard
import lekkit.scev.client.DisplayManager
import lekkit.scev.client.screen.owo.horizontalFlow
import lekkit.scev.client.screen.widget.IconButton
import lekkit.scev.main.ScalarEvolution
import lekkit.scev.menu.MachineMenu
import lekkit.scev.network.MachineInputPayload
import lekkit.scev.network.MachineResetPayload
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.components.Tooltip
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.player.Inventory
import net.neoforged.neoforge.network.PacketDistributor
import org.lwjgl.glfw.GLFW

/**
 * owo-ui port of the framebuffer / VM-input screen.
 *
 * Layout:
 *
 *   ┌── ScevSurfaces.PANEL ─────────────┐
 *   │  ┌── ScevSurfaces.INSET ────────┐ │
 *   │  │                              │ │
 *   │  │      framebuffer (4:3)       │ │
 *   │  │                              │ │
 *   │  └──────────────────────────────┘ │
 *   │  [ Power ]  [ Paste ]             │
 *   └───────────────────────────────────┘
 *
 * The framebuffer texture is painted through a custom owo [Surface]
 * ([framebufferSurface]) attached to a fixed-size cell. Going through
 * the Surface mechanism puts the texture into owo's draw order on top
 * of the parent INSET cell's `#464646` fill — no post-`super.render()`
 * hack needed. The cell's laid-out bounds also drive the input
 * handlers' hit-test via the [displayX]/[displayY]/[displayW]/[displayH]
 * mirror, which the `framebufferSurface` updates each frame.
 *
 * JEI exclusion is handled by `lekkit.scev.compat.jei.ScevJeiPlugin` —
 * the entire screen is marked as off-limits so JEI hides its overlay.
 *
 * **Held-key tracking.** We keep the set of HID keys we've pressed
 * (without release). On screen close / focus lost, every tracked key is
 * released — so if the player closes the GUI while holding shift, the
 * VM doesn't see shift as stuck down forever.
 */
class MachineScreen(menu: MachineMenu, inv: Inventory, title: Component) :
    ScevDisplayScreen<MachineMenu>(menu, inv, title) {

    private val keyPress: Consumer<Byte> = Consumer { hid ->
        PacketDistributor.sendToServer(MachineInputPayload.keyPress(hid))
    }
    private val keyRelease: Consumer<Byte> = Consumer { hid ->
        PacketDistributor.sendToServer(MachineInputPayload.keyRelease(hid))
    }

    private val heldKeys = HeldKeyTracker(keyPress, keyRelease)
    private val paster = ClipboardPaster(keyPress, keyRelease)

    /** 1.0 = render at native 640×480 resolution. Computed in
     *  [computeDisplaySize]. */
    private var displayScale: Float = 1.0f

    /** Top-left of the framebuffer rect in screen-space. Refreshed
     *  each frame from the Surface lambda's laid-out bounds; size
     *  comes from the base class's [displayW]/[displayH]. */
    private var displayX: Int = 0
    private var displayY: Int = 0

    /**
     * Custom owo [Surface] that blits the live framebuffer texture into
     * the component's bounds (or paints solid black if the VM is off).
     * Painting through a Surface puts the framebuffer into owo's normal
     * draw order, on top of the parent INSET cell's #464646 fill — no
     * post-super-render hack needed.
     */
    override val displaySurface: Surface = Surface { ctx, c ->
        val x = c.x()
        val y = c.y()
        val w = c.width()
        val h = c.height()
        // Mirror the laid-out rect for the input handlers + layout test.
        displayX = x; displayY = y

        val display = DisplayManager.get(menu.machineUuid)
        if (display == null) {
            ctx.fill(x, y, x + w, y + h, 0xFF000000.toInt())
            return@Surface
        }
        val tex = display.getOrUploadTexture()
        RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F)
        ctx.blit(tex, x, y, w, h, 0f, 0f, SCREEN_W, SCREEN_H, SCREEN_W, SCREEN_H)
    }

    init {
        // Provisional values — owo's actual layout overrides positioning,
        // but AbstractContainerScreen.init still wants something sane for
        // its leftPos/topPos calculation.
        imageWidth = SCREEN_W + 2 * SCREEN_MARGIN
        imageHeight = SCREEN_H + 2 * SCREEN_MARGIN + BUTTON_STRIP_H
    }

    /**
     * Target half-native (320×240 GUI pixels) by default — Minecraft's
     * own GUI scale setting then determines the actual on-screen pixel
     * size, the way vanilla inventories work. At GUI scale 4× on a 4K
     * monitor, half-native lands at 1280×960 actual px (~a third of the
     * screen) instead of 2560×1920 (most of it) like a 1.0 scale would.
     *
     * Auto-shrink only when the window genuinely can't fit the half-
     * native footprint plus chrome; never upscale past DEFAULT_SCALE.
     */
    override fun computeDisplaySize(): Pair<Int, Int> {
        val chromeW = 2 * (PANEL_BORDER + PANEL_PAD + INSET_PAD)
        val chromeH = chromeW + BUTTON_STRIP_H
        val availW = maxOf(SCREEN_W / 4, width - chromeW)
        val availH = maxOf(SCREEN_H / 4, height - chromeH)
        val scaleX = availW / SCREEN_W.toFloat()
        val scaleY = availH / SCREEN_H.toFloat()
        displayScale = minOf(DEFAULT_SCALE, minOf(scaleX, scaleY))
        return Pair(
            maxOf(1, Math.round(SCREEN_W * displayScale)),
            maxOf(1, Math.round(SCREEN_H * displayScale)),
        )
    }

    /**
     * Button row — Power resets the VM, Paste types the host clipboard.
     * Both are 12×12 sprite IconButtons (matching the existing
     * power_button.png 3-frame format) wrapped via owo's
     * wrapVanillaWidget. Text labels would compete with the framebuffer
     * for visual weight.
     */
    override fun buildBelowDisplay(panel: FlowLayout) {
        panel.child(horizontalFlow(Sizing.content(), Sizing.content()).apply {
            gap(BUTTON_GAP)
            child(Components.wrapVanillaWidget(makeIconButton(
                POWER_BUTTON_TEX, "button.scev.power", "tooltip.scev.power",
            ) { PacketDistributor.sendToServer(MachineResetPayload(false)) }))
            child(Components.wrapVanillaWidget(makeIconButton(
                PASTE_BUTTON_TEX, "button.scev.paste", "tooltip.scev.paste",
            ) { triggerPaste() }))
        })
    }

    /**
     * Build a 12×12 [IconButton] from a 42×14 sprite-sheet matching the
     * (idle, hover, pressed) layout of `power_button.png` (frames at
     * x=1, x=15, x=29, all at y=1). One static helper for all the
     * MachineScreen icon buttons; text labels live in the lang file
     * for narration / accessibility but aren't drawn.
     */
    private fun makeIconButton(
        sheet: ResourceLocation,
        narrationKey: String,
        tooltipKey: String,
        onClick: (IconButton) -> Unit,
    ): IconButton {
        val btn = IconButton(
            0, 0, ICON_BUTTON_SIZE, ICON_BUTTON_SIZE,
            sheet, ICON_SHEET_W, ICON_SHEET_H,
            15, 1,    // idle  (frame 1)
             1, 1,    // hover (frame 2)
            29, 1,    // pressed (frame 3)
            Component.translatable(narrationKey),
            { onClick(it) },
        )
        btn.tooltip = Tooltip.create(Component.translatable(tooltipKey))
        return btn
    }

    override fun containerTick() {
        super.containerTick()
        paster.tick()
    }

    /* ----- Input: keyboard ------------------------------------------------ */

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        if (keyCode == GLFW.GLFW_KEY_ESCAPE) return super.keyPressed(keyCode, scanCode, modifiers)

        // Ctrl+Shift+V (Cmd+Shift+V on macOS) pastes; plain Ctrl+V passes
        // through to the guest. The V key itself is consumed (not tracked
        // as held) so the guest doesn't see a stray 'v' before the paste.
        if (keyCode == GLFW.GLFW_KEY_V
            && (modifiers and GLFW.GLFW_MOD_SHIFT) != 0
            && (modifiers and (GLFW.GLFW_MOD_CONTROL or GLFW.GLFW_MOD_SUPER)) != 0
        ) {
            triggerPaste()
            return true
        }

        val hid = GlfwToHid.map(keyCode)
        if (hid != 0.toByte()) {
            heldKeys.press(hid)
            return true
        }
        return super.keyPressed(keyCode, scanCode, modifiers)
    }

    private fun triggerPaste() {
        var text = Minecraft.getInstance().keyboardHandler.clipboard
        if (text.isNullOrEmpty()) return
        if (text.length > MAX_PASTE_CHARS) text = text.substring(0, MAX_PASTE_CHARS)

        for (mod in MODIFIER_HIDS) {
            if (heldKeys.isHeld(mod)) {
                paster.queueRelease(mod)
            }
        }
        paster.queueText(text)
    }

    override fun keyReleased(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        if (keyCode == GLFW.GLFW_KEY_ESCAPE) return super.keyReleased(keyCode, scanCode, modifiers)

        val hid = GlfwToHid.map(keyCode)
        if (hid != 0.toByte()) {
            heldKeys.release(hid)
            return true
        }
        return super.keyReleased(keyCode, scanCode, modifiers)
    }

    /** Eat charTyped so typed text doesn't leak into debug overlays. */
    override fun charTyped(codePoint: Char, modifiers: Int): Boolean = true

    /* ----- Input: mouse --------------------------------------------------- */

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if (isInScreenRect(mouseX, mouseY)) {
            val hidBtn = mapMouseButton(button)
            if (hidBtn != 0.toByte()) {
                PacketDistributor.sendToServer(MachineInputPayload.mousePress(hidBtn))
                emitMousePlace(mouseX, mouseY)
                return true
            }
        }
        return super.mouseClicked(mouseX, mouseY, button)
    }

    override fun mouseReleased(mouseX: Double, mouseY: Double, button: Int): Boolean {
        val hidBtn = mapMouseButton(button)
        if (hidBtn != 0.toByte()) {
            PacketDistributor.sendToServer(MachineInputPayload.mouseRelease(hidBtn))
            return true
        }
        return super.mouseReleased(mouseX, mouseY, button)
    }

    override fun mouseDragged(mouseX: Double, mouseY: Double, button: Int, dx: Double, dy: Double): Boolean {
        if (isInScreenRect(mouseX, mouseY)) {
            emitMousePlace(mouseX, mouseY)
            return true
        }
        return super.mouseDragged(mouseX, mouseY, button, dx, dy)
    }

    override fun mouseScrolled(mouseX: Double, mouseY: Double, scrollX: Double, scrollY: Double): Boolean {
        if (isInScreenRect(mouseX, mouseY) && scrollY != 0.0) {
            val delta: Byte = if (scrollY > 0) 1 else -1
            PacketDistributor.sendToServer(MachineInputPayload.mouseScroll(delta))
            return true
        }
        return super.mouseScrolled(mouseX, mouseY, scrollX, scrollY)
    }

    override fun mouseMoved(mouseX: Double, mouseY: Double) {
        if (isInScreenRect(mouseX, mouseY)) emitMousePlace(mouseX, mouseY)
        super.mouseMoved(mouseX, mouseY)
    }

    /* ----- Lifecycle ------------------------------------------------------ */

    override fun removed() {
        paster.clear()
        heldKeys.releaseAll()
        super.removed()
    }

    /* ----- Helpers -------------------------------------------------------- */

    private fun isInScreenRect(mouseX: Double, mouseY: Double): Boolean =
        mouseX >= displayX && mouseX < displayX + displayW &&
        mouseY >= displayY && mouseY < displayY + displayH

    private fun emitMousePlace(mouseX: Double, mouseY: Double) {
        // Map GUI-space coords back to framebuffer pixels via inverse scale.
        // displayScale is the canonical scale; the actual displayBox dimensions
        // are equal up to rounding, so this is also the inverse of (displayW
        // / SCREEN_W) without the integer-rounding wobble.
        val scaleInv = if (displayScale == 0f) 1f else 1f / displayScale
        val x = maxOf(0, minOf(SCREEN_W - 1, ((mouseX - displayX) * scaleInv).toInt()))
        val y = maxOf(0, minOf(SCREEN_H - 1, ((mouseY - displayY) * scaleInv).toInt()))
        PacketDistributor.sendToServer(MachineInputPayload.mousePlace(x.toShort(), y.toShort()))
    }

    companion object {
        private const val SCREEN_W: Int = 640
        private const val SCREEN_H: Int = 480

        /**
         * Default GUI-space scale for the framebuffer. 0.5 → 320×240 GUI
         * pixels, which feels like a normal inventory window after MC's
         * own GUI-scale setting is applied (matches the look the 1.7.10
         * version had). Auto-shrinks when the window can't fit even this.
         */
        private const val DEFAULT_SCALE: Float = 0.5f

        // Outer scev PANEL has 1-px outline + 1-px bevel = 2 px on each
        // side. PANEL_PAD/INSET_PAD live in the base ScevDisplayScreen.
        private const val PANEL_BORDER: Int = 2

        // Reserved height for the button row in the auto-scale calculation
        // (button + gap + a bit of fudge so the panel's vertical sizing
        // doesn't have to be re-computed in two passes).
        private const val BUTTON_STRIP_H: Int = 24

        private const val BUTTON_GAP: Int = 4

        // Icon button sprite sheets (3 frames at x=1/15/29 of a 42×14 sheet,
        // see widget/power_button.png for the canonical layout — paste_button.png
        // matches the same shape).
        private const val ICON_BUTTON_SIZE: Int = 12
        private const val ICON_SHEET_W: Int = 42
        private const val ICON_SHEET_H: Int = 14
        private val POWER_BUTTON_TEX: ResourceLocation =
            ScalarEvolution.rl("textures/gui/widget/power_button.png")
        private val PASTE_BUTTON_TEX: ResourceLocation =
            ScalarEvolution.rl("textures/gui/widget/paste_button.png")

        private const val SCREEN_MARGIN: Int = 16   // legacy, kept for clarity in init()

        private const val MAX_PASTE_CHARS: Int = 1024

        private val MODIFIER_HIDS: ByteArray = byteArrayOf(
            HIDKeyboard.HID_KEY_LEFTCTRL, HIDKeyboard.HID_KEY_LEFTSHIFT,
            HIDKeyboard.HID_KEY_LEFTALT, HIDKeyboard.HID_KEY_LEFTMETA,
            HIDKeyboard.HID_KEY_RIGHTCTRL, HIDKeyboard.HID_KEY_RIGHTSHIFT,
            HIDKeyboard.HID_KEY_RIGHTALT, HIDKeyboard.HID_KEY_RIGHTMETA,
        )

        private fun mapMouseButton(glfwButton: Int): Byte = when (glfwButton) {
            GLFW.GLFW_MOUSE_BUTTON_LEFT   -> 1
            GLFW.GLFW_MOUSE_BUTTON_RIGHT  -> 2
            GLFW.GLFW_MOUSE_BUTTON_MIDDLE -> 4
            else -> 0
        }
    }
}
